### PR TITLE
[msvc] allow mozjs to be built using MSVC on windows

### PR DESCRIPTION
**Pull request by vvuk**
This change is [Reviewable](https://reviewable.io/reviews/servo/mozjs/82)

### CODE REVIEW

**Comment by Ms2ger** [comment]
Seems like some of those should go upstream too.

**Comment by vvuk** [comment]
WinNSPR isn't upstream yet :(
JS_PUBLIC_API stuff should though.

**Comment by metajack** [comment]
Is WinNSPR something we made then? I don't understand.

@ms2ger should sign off on this since he's in the middle of spidermonkey upgrade and these will need to be accounted for there too.

Build stuff looks fine to me.

---

Reviewed 6 of 7 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[build.rs, line 35 \[r2\]](https://reviewable.io:443/reviews/servo/mozjs/82#-KMfEa4QY2fwu-fwW8OL:-KMfEa4QY2fwu-fwW8OM:-1594204396) ([raw file](https://github.com/servo/mozjs/blob/eaf5c83c6feba4605f5c5d3b3478a97be9a39a7c/build.rs#L35)):*
> ```Rust
>             println!("cargo:rustc-link-lib=stdc++");
>         }
>         //println!("cargo:rustc-link-lib=mozjs-48a1");
> ```

Can we remove this commented out line?

---

*[makefile.cargo, line 46 \[r2\]](https://reviewable.io:443/reviews/servo/mozjs/82#-KMfEvsggzKvrKSq98ZR:-KMfEvsggzKvrKSq98ZS:1968339460) ([raw file](https://github.com/servo/mozjs/blob/eaf5c83c6feba4605f5c5d3b3478a97be9a39a7c/makefile.cargo#L46)):*
> ```
>         # Visual Studio build
> 	OUT_DIR:=$(shell cygpath "$(OUT_DIR)")
> 	CONFIGURE_FLAGS += --target=x86_64-pc-mingw32
> ```

The target is not MSVC?

---

*[makefile.cargo, line 51 \[r2\]](https://reviewable.io:443/reviews/servo/mozjs/82#-KMfFLjd2dp86_HcdosR:-KMfFLjd2dp86_HcdosS:289411242) ([raw file](https://github.com/servo/mozjs/blob/eaf5c83c6feba4605f5c5d3b3478a97be9a39a7c/makefile.cargo#L51)):*
> ```
> 	MOZ_TOOLS=/
> 
> 	ifneq (,$(wildcard c:/python27/python.exe))
> ```

Can we factor this into a branch for both windows targets so we aren't duplicating these lines?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/mozjs/82#-:-KMfFu5YlYekQhnPF3_J:2131442171)*

File: build.rs [hunk 1]
```diff
@@ -3,34 +3,55 @@
...
     let target = env::var("TARGET").unwrap();
-    let result = Command::new(env::var_os("MAKE")
-            .unwrap_or_else(|| OsStr::new(find_make()).to_os_string()))
+    let mut make = find_make();
+    // Put MOZTOOLS_PATH at the beginning of PATH if specified
+    if let Some(moztools) = env::var_os("MOZTOOLS_PATH") {
+        let path = env::var_os("PATH").unwrap();
+        let mut paths = env::split_paths(&path).collect::<Vec<_>>();
+        paths.push(PathBuf::from(moztools));
+        let new_path = env::join_paths(paths).unwrap();
```
Comment (Ms2ger) [inline build.rs]: Can we use `Iterator::chain` instead?

File: build.rs [hunk 2]
```diff
@@ -3,34 +3,55 @@
...
     assert!(result.success());
     println!("cargo:rustc-link-search=native={}/js/src", out_dir);
-    println!("cargo:rustc-link-lib=static=js_static");
     if target.contains("windows") {
         println!("cargo:rustc-link-lib=winmm");
         println!("cargo:rustc-link-lib=psapi");
+        if target.contains("gnu") {
+            println!("cargo:rustc-link-lib=stdc++");
+        }
+        //println!("cargo:rustc-link-lib=mozjs-48a1");
```
Comment (Ms2ger) [inline build.rs]: Rm

File: build.rs [hunk 3]
```diff
@@ -3,34 +3,55 @@
...
+        println!("cargo:rustc-link-lib=static=js_static");
+    } else {
+        println!("cargo:rustc-link-lib=stdc++");
+        println!("cargo:rustc-link-lib=static=js_static");
```
Comment (Ms2ger) [inline build.rs]: Let's move this back out of the if/else

**Comment by Ms2ger** [comment]
You'll need to update `etc/patches`

**Comment by vvuk** [comment]
Review status: 2 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[makefile.cargo, line 46 \[r2\]](https://reviewable.io:443/reviews/servo/mozjs/82#-KMfEvsggzKvrKSq98ZR:-KOLz7HdWnvt7gFnakWa:364076152) ([raw file](https://github.com/servo/mozjs/blob/eaf5c83c6feba4605f5c5d3b3478a97be9a39a7c/makefile.cargo#L46)):*
<details><summary><i>Previously, metajack (Jack Moffitt) wrote…</i></summary>
> The target is not MSVC?
</details>
Not for gecko -- it uses the target as written and does internal stuff to figure out what compiler is being used.  Gecko's build system is a mess, just nod and smile :|

---

*[makefile.cargo, line 51 \[r2\]](https://reviewable.io:443/reviews/servo/mozjs/82#-KMfFLjd2dp86_HcdosR:-KOLzF4br_lKmRzGLC0S:676635175) ([raw file](https://github.com/servo/mozjs/blob/eaf5c83c6feba4605f5c5d3b3478a97be9a39a7c/makefile.cargo#L51)):*
<details><summary><i>Previously, metajack (Jack Moffitt) wrote…</i></summary>
> Can we factor this into a branch for both windows targets so we aren't duplicating these lines?
</details>
Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/mozjs/82)*

**Comment by metajack** [comment]
New changes look good to me. ms2ger, anything left you are waiting on?

---

Reviewed 1 of 5 files at r3, 2 of 2 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/mozjs/82#-:-KOM2arPWjbqgVe3u6Y0:2132499614)*

**Comment by vvuk** [comment]
> New changes look good to me. ms2ger, anything left you are waiting on?

I'm not sure how to make `Iterator::chain` happen btw, would be happy to if someone can guide me :)

**Comment by larsbergstrom** [comment]
Success: https://travis-ci.org/servo/mozjs/builds/150001941

**Pull request merged**